### ISSUE TITLE
[css-anchor-position-1] Behavior with transforms and offset-path

### ISSUE THREAD

**Issue by kizu** [w3c/csswg-drafts#8584]
<details>
<summary>Disclaimer and additional links</summary>

I'm submitting my feedback following my experiments with the current implementation of anchor positioning in Chrome Canary.

I wrote an article about my experiments, but decided to fill most of my feedback as separate issues here.

A quick summary of related links:

- [anchor positioning specs](https://drafts.csswg.org/css-anchor-position-1/)
- [my article with the experiments](https://kizu.dev/anchor-positioning-experiments/)
</details>

Right now, the positions for the anchors do not take into account things like transforms, `offset-path` (`position: sticky`, which is covered by https://github.com/w3c/csswg-drafts/issues/8448).

In practice, there were cases in my experiments where I wanted to use those — transforms and `offset-path` for the animations/transitions (see the animations for the [“four quadrants”](https://kizu.dev/anchor-positioning-experiments/#the-technique-itself) in my article — I want to attach things that move, but because transforms or offset-paths are not used, I have to animate things via `top` and `left`).

Currently, specs do not mention any of these cases — if those are intended to not be supported, the specs should be clarified, but I would suggest looking into providing an option to somehow follow those.

I understand that for the transforms and offset-path things are probably much more complicated — but I foresee a lot of cases where developers would want an option to follow those. I wonder if this could be something like `anchor-transform` or whatever, which could be optionally used with `transform` on the anchored element? And maybe something similar with `offset-path` (though, maybe it could be similar to `position: sticky` in that it is easier to describe how it could work by default?)

**Comment by xiaochengh** [comment]
It should definitely work with `position: sticky`, but we haven't done the spec work yet: #8448

Would you mind splitting this bug by moving the stickypos related stuff to #8448 and leaving the remaining here?

**Comment by kizu** [comment]
(copied with modifications from the OpenUI discord)

I thought about a case where we'd want to follow the transform of the target's container: https://codepen.io/kizu/pen/WNLYRvx

Basically, what if we want to attach a popover to a button inside a rotated, or otherwise transformed container?

Here is what we get right now in this example (in Canary with experimental platform features flag due to anchor positioning), when the element is outside of the rotated container (inside the rotated one things are obviously ok), or when it is yoinked into the topmost layer via popover API:

<img width="306" alt="A screenshot of an example with a button and a popover, where the button is placed in a rotated container, with a popover attached to it, but the popover itself is not rotated, so it is “detached” from the button" src="https://github.com/w3c/csswg-drafts/assets/177485/cc3915d6-cb42-43c4-91cb-8ac164036000">

This can also be a case for when someone implements a sliding side-panel, with sliding animated via a transform, but then what if the panel would contain a `defaultOpen` popover that should be attached to some element via anchor positioning? With the way things work now, that element would exist completely outside of the transformed space, and to replicate what we want to achieve, we'd have to do _a lot_ of work by manually transforming the popover.

For regular abspos, we could argue, that the idea was that we now could place the popover near the element, so we would expect the behavior as in the third container in the codepen:

<img width="345" alt="A screenshot of a rotated container, a button inside and an abspos element attached to the button, following the rotation of the container." src="https://github.com/w3c/csswg-drafts/assets/177485/22f8f6f9-5f07-4e1c-a14a-7db6e5cb8d47">

But then, when we have a Popover API — the element is reparented to the topmost layer, thus loses that transformation context.

Should we be able to either somehow follow a transformation of a selected element, or be able to disable putting the element in the topmost layer (which, I imagine, can be problematic on its own?), or something else?

**Comment by xiaochengh** [comment]
It seems doable to me if we want to additionally apply additional transforms to the anchored element after layout. There are some issues with the transform origin, but they may be fixable.

The issue is that it won't work well with position fallback, because it's post-layout. And I don't see any way to make position fallback work with general transforms.

**Comment by kizu** [comment]
I think it might be reasonable to lose the ability to apply fallbacks when following the transforms. Though, I wonder if this is something that might benefit from a property/keyword somewhere that controls this? Could there be cases where there'd be a transform which we don't want to follow? Something like “neutral” transforms, like the initial state ones (`scale(1)`, `translateZ(0)` etc) come to mind, where an element could have it, but the developers would not want this to interrupt the fallback mechanism.

**Comment by xiaochengh** [comment]
> Could there be cases where there'd be a transform which we don't want to follow?

I think this might be the common case. If the style doesn't explicitly have a transform, then there shouldn't be any transform.

Exceptions might only be cases similar to `<selectlist>`, which look broken if the popover doesn't follow the transform by default.

**Comment by bfgeek** [comment]
Right - we don't want to allow a layout dependency on transforms unfortunately. Lots of things are accelerated for users, and this would force a bunch of things onto a slow path.

**Comment by yisibl** [comment]
There doesn't seem to be any discussion of offset-path here, do I need to open a separate issue?

Test in Chrome 127:
https://codepen.io/yisi/pen/eYaJopE

https://github.com/w3c/csswg-drafts/assets/2784308/840b1cdb-f3b2-4240-b2e0-9a08d82a633c

**Comment by jonrimmer** [comment]
Many web apps with a "canvas" UI use transforms as a key part of their rendering. I.e. `transform()` to handle panning and `scale()` to handle zooming. Anything built using [react-flow](https://reactflow.dev/) for example. Adding tooltips, popovers and context menus on canvas items is an common requirement for these apps, but if I read this correctly, it will be impossible to use anchor positioning for these kinds of use-cases? If so, that seems like a big gap.

**Comment by jonrimmer** [comment]
It's perhaps also worth nothing that the most popular JS library for achieving anchored positioning, [Floating UI](https://floating-ui.com/) (the successor to Popper.js), _does_ take CSS transforms into account, so as it stands anchor-positioning will not be adequate to replace it.

**Comment by emilio** [comment]
It is possible to apply offsets and scales in these cases, you just need it to explicitly do it on the anchored element as well.

The issue with accounting for transforms is that it'd force the transform animations on such elements (and any of their ancestors) to not be off-main-thread, _and_ affect layout, which is a no-no.

**Comment by AlbertMarashi** [comment]
@jonrimmer indeed I am here because my app is using canvas features like translate and scale, and having anchored elements that could scale with both would be great

currently we are doing all of this inside a animation frame loop

**Comment by michaelinva** [comment]
I don't think it is needed for the anchored element to transform with the target element's own transformations. We could always just apply the transformation to that target element too with javascript or put them both under the same DOM parent like already mentioned here. Chrome and Safari implement it exactly like that; no need to overthink it.

**Comment by mstensho** [comment]
So, what should the spec say?

Current status:

https://drafts.csswg.org/css-anchor-position-1/#determining-position
> Post-layout effects, such as [transform](https://drafts.csswg.org/css-transforms-1/#propdef-transform), do not affect the [anchor box’s](https://drafts.csswg.org/css-anchor-position-1/#anchor-box) position.
> Note: Allowing an anchor to opt into including the effects of [transform](https://drafts.csswg.org/css-transforms-1/#propdef-transform) or similar properties might be allowed in the future.

Then, about taking scrolling into account: https://drafts.csswg.org/css-anchor-position-1/#issue-141fd32f
>  Transforms have the same issue as scrolling, so Anchor Positioning similarly doesn’t pay attention to them normally. Can we go ahead and incorporate the effects of transforms here?

And: https://drafts.csswg.org/css-anchor-position-1/#issue-a8569080
> Similar to [remembered scroll offset](https://drafts.csswg.org/css-anchor-position-1/#remembered-scroll-offset), can we pay attention to transforms on the [default anchor element](https://drafts.csswg.org/css-anchor-position-1/#default-anchor-element)?

Do we want to take an anchor's transforms into account, and how far would we like to go? Only honor those on the default anchor when at an anchor recalculation point? Or more? We don't want to allow transforms to affect layout, ever, I assume, not even as an opt-in? Because that means main-thread work, and no compositing.

Otherwise we could do funny things like this (video below):

```html
<!DOCTYPE html>
<style>
  .transform {
    transition: 2s transform;
  }
  .transform:hover {
    transform: rotate(500deg) scale(0.5) translateX(200px);
  }
  .dot {
    position: absolute;
    position-anchor: --a;
    width: 20px;
    height: 20px;
    border-radius: 50%;
    background: hotpink;
  }
</style>
<div style="width:fit-content; margin:100px;">
  <div class="transform" style="display:inline-block; anchor-name:--a; width:350px; height:350px; background:cyan;">
    Hover me. Whee!
  </div>
</div>
<div class="dot" style="position-area:right;"></div>
<div class="dot" style="position-area:left;"></div>
<div class="dot" style="position-area:top;"></div>
<div class="dot" style="position-area:bottom;"></div>
```

https://github.com/user-attachments/assets/a8ce8ef9-3f53-4e17-b49a-871a9876fc4d

And by taking an anchor's transforms into account, use its bounding box after transforms have been applied?

**Comment by tabatkins** [comment]
> Do we want to take an anchor's transforms into account, and how far would we like to go? Only honor those on the default anchor when at an anchor recalculation point? Or more? We don't want to allow transforms to affect layout, ever, I assume, not even as an opt-in? Because that means main-thread work, and no compositing.

Right, the potential for *layout* effects is the obvious problem here.

So just to review how we handle scroll:

* at anchor recalculation points, we calculate all anchor references based on their actual scrolled position. (So `anchor()` resolves to the scroll-accurate offset, and you can get accurate results from `max(anchor(top), anchor(--foo top))`/etc.) This *is* capable of changing the the size of the positioned element.
* at all times, we track the default anchor's scroll offset, and when it differs from the remembered scroll offset from the last anchor recalculation point, we apply the difference as a translate to the positioned element. (So `anchor()`/etc do *not* respond to this difference, and the `max()` example might no longer be correct, etc.) Since it's a transform, this is *not* capable of changing the size of the positioned element; it just shifts it.

The live-responding transform is a little clumsy, since it can break constraints you write into your offsets. It's a reasonable compromise, tho, for the benefit of being able to "stick to" the default anchor reliably.

If we want transforms to work similarly, we need to adopt similar constraints.
* at anchor recalculation points, we calculate all anchors based on their actual *transformed* position (using the axis-aligned 2d bounding box of the transformed anchor).
* at all times, we track the default anchor's transform. We need to track some *point* on the default anchor (I'll discuss which point later) - when its position differs from its position at the last anchor recalculation point, we apply the difference as a translate.

Which point, tho? Translates don't really care which we use, but rotates and scales definitely do. And is the point on the anchor itself, or on its transform bounding box? I think the right answer is that it's a point on the bounding box. But which point? I think the "right" answer is going to depend on some other properties.

* We should have a property (`anchor-transform`?) that lets you specify this point explicitly, so you can do whatever you want.
* Otherwise (if the property is `auto`), if the positioned element is using `position-area`, we can infer a reasonable point based on the area:
	* If the area has exactly one corner touching one of the anchor's corners, use that corner. (Numbering the area sections like a phone keypad, areas like "1" or "1245".)
	* Otherwise, if the area has exactly one edge touching one of the anchor's sides in an axis, use the center of that side. (Areas like "12", "123", "25", or "123456".) (It can either touch both sides in the other axis, like "25", or neither in the other axis, like "123".)
	* Otherwise, use the center of the bounding box. (Areas like "5" or "123456789".)
* Otherwise, use the center of the bounding box. (This will make it respond to translations, but *not* the anchor rotating or scaling itself at all. That's intentional, since we can't guess at what part we're intending to anchor to.)

---------

The unstated next question is whether it's possible to do this by default at this point, or if we need to continue ignoring transforms by default and key it to the property that controls what point we're tracking (like `anchor-transform: none` as the initial value, and `anchor-transform: auto`/etc would turn it on).

**Comment by mstensho** [comment]
I suggest that we focus on dealing with anchor recalculation points first, and then figure out any opt-in mechanism for "live" transforms later.

It seems pretty obvious to me how this (at recalculation points) is supposed to work, so, although the spec needs to be updated, I can go ahead and land something.

**Comment by tabatkins** [comment]
Right. So just to confirm, so I can edit accordingly: in the section where we talk about handling scroll at recaluculation points, we also say that we pay attention to transforms, and use the transform bounding box 
 of the anchor (in the coordinate space of the positioned element's containing block) instead of the anchor's actual box.

I note this does interact a bit with the request to use different boxes from the anchor (right now we always use the anchor's border box edge). Not sure how we'd handle using padding/content box, unless maybe by applying the same transforms to that reduced box and then finding the bounding box of *that*.

Anyway, something to solve for later; paying attention to transforms on the anchor is definitely more important than caring about the content box of an anchor.

**Comment by tabatkins** [comment]
> When an anchor is resized or moved, that should obviously affect any anchored elements, so any such bounding box that we calculated at the previous anchor recalculation point needs to be updated somehow. But transforms may have changed in the meantime as well. What to do?

Ah, took me a sec to realize what the question was here. We track layout changes "live", but only track scroll offset changes at the specific recalculation points, and the question is whether transform changes should act like layout or like scroll.

I agree it would be weird if both the layout and transform changes, but we don't update the transform adjustment, because the anchor would be off in a weird way. 

But what if *only* the transform updates? That is meant to be hardware-accelerateable, like scrolling, so tracking it live would cause the same issues as tracking scrolling - it might be a frame or two late, which looks really bad when the positioned element is trying to "stick" to the anchor.

I'm not sure what the right answer is here. :/

**Comment by chrishtr** [comment]
> I'm not sure what the right answer is here. :/

I think we should treat transform the same as layout-affecting properties--if it changes, then schedule a re-render. This will get behind by a frame or so during a threaded animation of transform, but that's better than breaking when transform+layout changes at the same time.

**Comment by fantasai** [comment]
One question that came up in the call was whether this should affect all transforms, or only non-`transform` transforms (which are more limited).

**Comment by tabatkins** [comment]
They're not meaningfully limited. `translate`/`rotate`/`scale` can do 2d *or* 3d transforms, and `offset-path` can do a combination of 2d translate and rotate. The only real extra power that `transform` gives is the skew functions, and arbitrary matrix, but those are not meaningfully different for these purposes.

(If it was a "2d vs 3d" distinction there *might* have been an argument, I think, but since that ship's already sailed, I don't see any reason to distinguish the various transform sources.)

**Comment by mstensho** [comment]
Does this mean that my example in https://github.com/w3c/csswg-drafts/issues/8584#issuecomment-2879321841 should work, including resizing of the gray rectangles in the corners?

**Comment by chrishtr** [comment]
> Does this mean that my example in [#8584 (comment)](https://github.com/w3c/csswg-drafts/issues/8584#issuecomment-2879321841) should work, including resizing of the gray rectangles in the corners?

Yes, I think so. Except that the anchored element might fall behind during the transition animation by a frame or so.

**Comment by tabatkins** [comment]
Right. Except for the possibility of frame slippage, it should work exactly the same as just animating the width/height/position of a box (matching the size/position of the bounding box) and anchoring to that instead.

**Comment by john-melin** [comment]
Many virtual scroll solutions use transform to scroll recycled elements. Without a way to allow anchor-position to follow those transforms creating native tooltips using popover="hint" wouldn't be possible.

So although slower if default it should be possible to opt-in to following transform.

**Comment by tabatkins** [comment]
All right, transform text is in the spec now, in https://drafts.csswg.org/css-anchor-position/#anchor-position-size.

As far as @fantasai and I could tell, this only really needed a few sentences to define. Are there any details we missed?

**Comment by devongovett** [comment]
Should there be a way to opt-out of this? I have an example of a button that has a shake animation (a rotation transform) when an error occurs, and this is causing an attached tooltip to also shake. Ideally the tooltip would stay fixed while the button animation occurred.

**Comment by kizu** [comment]
@devongovett, I think that's a really nice use case for an opt-out for sure! Can you open a separate issue about it?

As a workaround, separating the anchor from the transformed element would help (so the tooltip is attached to the outer wrapper, and only the inner element is transformed), but this is for sure not something we want authors to have to do, and it would be great to have a way to opt out for sure.

**Comment by devongovett** [comment]
@kizu #13782
